pkg/ccl/changefeedccl/alter_changefeed_stmt.go

```diff
@@ -317,8 +317,18 @@ func generateNewTargets(
 
 	for _, target := range AllTargets(prevDetails) {
 		k := targetKey{TableID: target.TableID, FamilyName: tree.Name(target.FamilyName)}
+		desc := descResolver.DescByID[target.TableID].(catalog.TableDescriptor)
+		tbName, err := getQualifiedTableNameObj(ctx, p.ExecCfg(), p.ExtendedEvalContext().Txn, desc)
+		if err != nil {
+			return nil, nil, hlc.Timestamp{}, err
+		}
+		tablePattern, err := tbName.NormalizeTablePattern()
+		if err != nil {
+			return nil, nil, hlc.Timestamp{}, err
+
+		}
```

Comment:
I can see why we are doing what we're doing here... But the conversion from current TargetSpecification -> ChangefeedTarget looses StatementTimeName I think.

Consider what happens when you created changefeed on table "foo", but then, later renamed it to "bar".  Wouldn't you loose "foo" statement time name?  Even ignoring the fact that in the meantime you could have created another foo table.  

I think that's the problem @HonoreDB mentioned yesterday.

Ignoring for a second whether we should allow such operation in the first place, I think we need to deal with this rename problem for now.

I think we can probably do something about this.  Let's add a "resolved flavor" to ChangefeedTarget -- somewhat similar to "resolve name" and "unresolved name".

In `pkg/sql/sem/tree/changefeed.go` let's add an un-exported field to `ChangefeedTarget` containing resolved jobspb.ChangefeedTargetSpecification... We almost certainly won't be able to add this field directly -- i.e. using jobspb inside tree package is probably not a good idea... but still, we can probably do:

```
type ChangefeedTarget struct {
	TableName  TablePattern
	FamilyName Name
  resolvedTarget interface{}
}

// MakeResolvedChangefeedTarget construct changefeed target
// that has already been resolved.
func MakeResolvedChangefeedTarget(resolved interface) ChangefeedTarget {
   return ChangefeedTarget{resolvedTarget: resolved}
}

func (ct *ChangefeedTarget) ResolvedSpec() interface {
  return ct.resolved
}
```

I don't think we need anything more in the tree package.  In particular, you don't need
to worry about being able to format that "resolved" field as a string because this interface is pretty much added for internal communication between alter statement and the rest of the changefeed code.  

Alter changefeed, instead of doing the above code, would construct MakeResolvedChangefeedTarget using existing "target".
What else needs to change? Probably just the `getTargetsAndTables` in changefeed_statement so that it looks to see if target has non nil ResolvedSpec and just uses it.

What do you think?
@HonoreDB does that sound reasonable to you?

Reply:
That makes sense to me, yeah--we're telling changefeed_stmt to reuse an old name resolution instead of re-doing it. If it feels too weird to mess with tree, an equivalent thing could be to build a `originalSpecs` map of tree.ChangefeedTarget to ChangefeedTargetSpecification and pass it around.

Reply:
Yeah, if there is a way to avoid that -- great... It's just we still need to construct ChangefeedStatement to pass to the rest of the code.  I think it's important to use the "original" logic in changefeed_statement wrt to validation, job record creation, etc.  So, one way or another, this information needs to me in the form of a ChangefeedStatement.
Now, in the scheduled backup (and elsewhere), I used a trick where I introduced annotated statement:
```
// annotatedBackupStatement is a tree.Backup, optionally
// annotated with the scheduling information.
type annotatedBackupStatement struct {
	*tree.Backup
	*jobs.CreatedByInfo
}
```
And back would see if the tree.Backup is an annotated statement (i.e. the call is made from the scheduled jobs system), and do something else.  I suppose we can choose to maintain original targets by passing e.g.
```
type alterChangefeedStmt struct {
   *tree.ChangefeedStatement
   originalSpecs ...
}
```
And do something appropriate with this.